### PR TITLE
[ENHANCEMENT] Generate component-tests into `tests/integration` by default

### LINKED ISSUE #4178: Update blueprints for components tests so they go under tests/integrations

### PR DESCRIPTION
**Pull request by trabus**
resolves #4178
By default component-tests now generate into `tests/integration`, with a new `--test-type` option allowing you to generate as a unit test intstead. The `--test-type` option can only be `unit` or `integration` and has the aliases `-unit`, `-u`, `-integration`, and `-i`.
```
ember g component-test foo-bar -unit
```

This PR also adds support for using an array of possible values as a command option type, accepting only the values in the array as valid arguments (essentially an enum type). For example, this is what the option for `test-type` looks like:

```
  availableOptions: [
    {
      name: 'test-type',
      type: ['integration', 'unit'],
      default: 'integration',
      aliases:[
        { 'i': 'integration'},
        { 'u': 'unit'},
        { 'integration': 'integration' },
        { 'unit': 'unit' }
      ]
    }
  ]
```

### CODE REVIEW

**Comment by trabus** [comment]
@ef4 @stefanpenner ready for review!

**Comment by kiwiupover** [comment]
thanks @trabus for your work here.

File: blueprints/component-test/index.js [hunk 1]
```diff
@@ -4,13 +4,30 @@ var path          = require('path');
...
+        { 'integration': 'integration' },
+        { 'unit': 'unit' }
+      ]
+    }
+  ],
 
   fileMapTokens: function() {
     return {
+      __testType__: function(options) {
+        return options.locals.testType || 'integration';
```
Comment (stefanpenner) [inline blueprints/component-test/index.js]: it seems like we have the "default" value specified in two locations. Should be merge to a single source of truth?

**Comment by stefanpenner** [comment]
left one comment, it may be a valid concern. Otherwise though, LGTM

File: blueprints/component-test/index.js [hunk 2]
```diff
@@ -22,12 +39,21 @@ module.exports = {
...
+      testTypeDefinition = "// Specify the other units that are required for this test" + 
+        EOL + "  // needs: ['component:foo', 'helper:bar']," + EOL + "  unit: true";
+    }
+    
     return {
       path: getPathOption(options),
+      testType: options.testType ,
       componentPathName: componentPathName,
+      testTypeDefinition: testTypeDefinition,
       friendlyTestDescription: testInfo.description(options.entity.name, "Unit", "Component")
```
Comment (pangratz) [inline blueprints/component-test/index.js]: I think we should use different `friendlyTestDescription`s for each case...

**Comment by ef4** [comment]
:+1:

File: blueprints/component-test/index.js (hunk 1 above)
Comment (trabus) [inline blueprints/component-test/index.js]: I'd like to clean this up so it's not necessary, but there are situations where [addon developers could override locals](https://github.com/ember-cli/ember-cli/issues/4183), making testType unavailable.

File: blueprints/component-test/index.js (hunk 2 above)
Comment (trabus) [inline blueprints/component-test/index.js]: Good catch. I'll put that in asap.

**Comment by pangratz** [comment]
Wonderful :heart_eyes: 

The sample test case for `integration` needs to be a different case though, since using `this.subject` [will throw an error](http://git.io/vIPao) within an integration test ...

**Comment by trabus** [comment]
@pangratz another good catch. on it now.

**Comment by trabus** [comment]
Actually, this touches on #4201, but I think I'll leave the asserts in and deal with them in a PR for that issue.

**Comment by trabus** [comment]
@pangratz Let me know if this is any better. Added some notes for setting up integration tests to the output.

File: blueprints/component-test/index.js [hunk 3]
```diff
@@ -22,13 +39,41 @@ module.exports = {
...
     var dasherizedModuleName = stringUtil.dasherize(options.entity.name);
     var componentPathName = dasherizedModuleName;
-    if(options.pod && options.path !== 'components' && options.path !== '') {
+    var testTypeDefinition = "integration: true";
+    var friendlyTestDescription = testInfo.description(options.entity.name, "Integration", "Component");
+    var testContent = "assert.expect(1);" + EOL + EOL +
+      "  // Set any properties with this.set('myProperty', 'value');" + EOL +
+      "  // Handle any actions with this.on('myAction', function(val) { ... });" + EOL + EOL +
+      "  // Provide a template (string or precompiled) for this.render()" + EOL +
+      "  this.render();" + EOL + EOL +
```
Comment (pangratz) [inline blueprints/component-test/index.js]: Something needs to be passed to `render`, as otherwise an [error is thrown](http://git.io/vIPAi). The problem is now that the `ember-template-compiler.js` needs to be present ... :cry:

I plan to open a PR for including `htmlbars-inline-precompile` as default, so this line will eventually be changed to:

``` js
this.render(hbs`
  {{<%= dasherizedModuleName %>}}
`);
```

Not sure what to do now? Sadly the current version of the integration test will throw an error, but maybe that's ok and this should be fixed before a new release is cut..

**Comment by pangratz** [comment]
@trabus besides my comment, this looks good!

File: blueprints/component-test/index.js (hunk 3 above)
Comment (trabus) [inline blueprints/component-test/index.js]: Maybe a failing test to start isn't so bad?

**Comment by trabus** [comment]
@pangratz Thanks for the review(s)!

@stefanpenner The changes I just had to make to have the testContent templated absolutely underscore the justification for implementing #3815.

File: blueprints/component-test/index.js (hunk 3 above)
Comment (pangratz) [inline blueprints/component-test/index.js]: > Maybe a failing test to start isn't so bad?

In some circumstances this might not be so bad, but the problem is just that fixing the failing test requires adding the `ember-template-compiler.js` to the `Brocfile` – or installing the `htmlbars-inline-precompile`.

Also, this would be the first time a generated test would fail ...

Comment (trabus) [inline blueprints/component-test/index.js]: Can we just get away with adding an empty `'<span></span>'` or something?

Comment (pangratz) [inline blueprints/component-test/index.js]: Unfortunately not since `ember-test-helpers` tries to compile a passed string via [`Ember.Handlebars.precompile()`](http://git.io/vIXtj) which will :boom: ...

Comment (pangratz) [inline blueprints/component-test/index.js]: @stefanpenner https://github.com/babel/ember-cli-babel/pull/42 is currently blocking https://github.com/pangratz/ember-cli-htmlbars-inline-precompile/issues/7. When those are resolved, I will open a PR for using `htmlbars-inline-precompile` inside component integration tests, as suggested in https://github.com/pangratz/ember-cli-htmlbars-inline-precompile/issues/7#issuecomment-103967882.

**Comment by blimmer** [comment]
I just implemented similar functionality to this in [ember-cli-mocha](https://github.com/switchfly/ember-cli-mocha/pull/57). @rwjblue asked me to do a PR to ember-cli to match that, but it's similar to this PR, so I'm hoping to get some feedback before I proceed.

Main Similarities:
* Introduces tests/integration/components/*-test.js

Main Differences:
* Uses [ember-cli-htmlbars-inline-precompile](https://github.com/pangratz/ember-cli-htmlbars-inline-precompile) to precompile the `render` statement.
* Does not change the default behavior when generating a component (a unit test is generated automatically and you have to explicitly generate component integration tests with `ember g component-test -i`)

What I'm wondering is:
1. Do we want to generate an integration test by default instead of a unit test?
2. Is there any issue adding a new dependency on ember-cli-htmlbars-inline-precompile?

**Comment by rwjblue** [comment]
I would like to use integration by default, and use the inline precompiler for those integration tests. 

I haven't had a bunch of time to review this PR, though so am not 100% sure what it does....

**Comment by trabus** [comment]
@blimmer I can bring this PR inline with yours regarding the ember-cli-handlebars-inline-precompile, but I think the default behavior of this PR is preferred (generating integration by default).

**Comment by blimmer** [comment]
Sounds good. I can make a PR against ember-cli-mocha to bring the two back in sync (integration test by default and matching the arguments defined here).

**Comment by trabus** [comment]
> Sounds good. I can make a PR against ember-cli-mocha to bring the two back in sync (integration test by default and matching the arguments defined here).

@blimmer  :+1:
Thanks!

File: blueprints/component-test/index.js [hunk 4]
```diff
@@ -22,13 +39,43 @@ module.exports = {
...
     var dasherizedModuleName = stringUtil.dasherize(options.entity.name);
     var componentPathName = dasherizedModuleName;
-    if(options.pod && options.path !== 'components' && options.path !== '') {
+    var testImports = EOL + "import hbs from 'htmlbars-inline-precompile';" + EOL;
+    var testOptions = "integration: true";
+    var friendlyTestDescription = testInfo.description(options.entity.name, "Integration", "Component");
+    var testContent = "assert.expect(1);" + EOL + EOL +
+      "  // Set any properties with this.set('myProperty', 'value');" + EOL +
+      "  // Handle any actions with this.on('myAction', function(val) { ... });" + EOL + EOL +
+      "  this.render(hbs`{{" + dasherizedModuleName + "}}`);" + EOL + EOL +
```
Comment (pangratz) [inline blueprints/component-test/index.js]: I wonder if this should be changed to show how components with a template block can be tested now? Maybe something like this:

``` js
"this.render(hbs`" + EOL
"  {{#" + dasherizedModuleName + "}}" + EOL
"    template block text" + EOL
"  {{/" + dasherizedModuleName + "}}" + EOL
"`);" + EOL + EOL +
"assert.equals(this.$().text().trim(), 'template block text');"
```

What do you think?

**Comment by rwjblue** [comment]
:+1:

File: blueprints/component-test/index.js (hunk 4 above)
Comment (trabus) [inline blueprints/component-test/index.js]: Shoot, this didn't make it in, but now that it's in master, we can just do another PR for clarifiying block usage.

Comment (rwjblue) [inline blueprints/component-test/index.js]: Yeah, sorry about that.  Thanks for cleaning it up.

Comment (trabus) [inline blueprints/component-test/index.js]: So would it be preferred to default with a template block, or just have it as an example?

Comment (trabus) [inline blueprints/component-test/index.js]: We could put both plain usage and block usage in, and leave it to the user to delete the part they're not using. I believe having both tests would pass on initial creation.

This is what I currently have:
```
      "  // Set any properties with this.set('myProperty', 'value');" + EOL +
      "  // Handle any actions with this.on('myAction', function(val) { ... });" + EOL + EOL +
      "  this.render(hbs`{{" + dasherizedModuleName + "}}`);" + EOL + EOL +
      "  assert.equal(this.$().text(), '')" + EOL + EOL +
      "  // Template block usage:"
      "  this.render(hbs`" + EOL
      "    {{#" + dasherizedModuleName + "}}" + EOL
      "      template block text" + EOL
      "    {{/" + dasherizedModuleName + "}}" + EOL
      "  `);" + EOL + EOL +
      "  assert.equals(this.$().text().trim(), 'template block text');"
```

Comment (rwjblue) [inline blueprints/component-test/index.js]: Can you also delete the test that uses `component._state` which would resolve https://github.com/ember-cli/ember-cli/issues/4201?

Comment (trabus) [inline blueprints/component-test/index.js]: Will do. Replacing with:
```
      testContent = "assert.expect(1);" + EOL + EOL +
        "  // Creates the component instance" + EOL +
        "  var component = this.subject();" + EOL +
        "  // Renders the component to the page" + EOL +
        "  this.render();" + EOL +
        "  assert.equal(this.$().text(), '');";
```
